packages/react-router/src/qss.ts

```diff
@@ -33,18 +52,33 @@ function toValue(mix) {
   return +str * 0 === 0 && +str + '' === str ? +str : str
 }
 
+/**
+ * Decodes a query string into an object.
+ * @param str - The query string to decode.
+ * @param [pfx] - An optional prefix to filter out from the query string.
+ * @returns The decoded key-value pairs in an object format.
+ * @example
+ * // Example input: decode("token=foo&key=value")
+ * // Expected output: { "token": "foo", "key": "value" }
+ */
 export function decode(str, pfx?: string) {
   let tmp, k
   const out = {},
     arr = (pfx ? str.substr(pfx.length) : str).split('&')
 
   while ((tmp = arr.shift())) {
-    tmp = tmp.split('=')
-    k = tmp.shift()
-    if (out[k] !== void 0) {
-      out[k] = [].concat(out[k], toValue(tmp.shift()))
+    const equalIndex = tmp.indexOf('=')
+    if (equalIndex !== -1) {
+      k = tmp.slice(0, equalIndex)
+      const value = tmp.slice(equalIndex + 1)
+      if (out[k] !== void 0) {
+        out[k] = [].concat(out[k], toValue(value))
+      } else {
+        out[k] = toValue(value)
+      }
     } else {
-      out[k] = toValue(tmp.shift())
+      k = tmp
```

Comment:
Do we need this?

What happens is that if the key is empty ("?token"), an empty string is assigned to the key in the `out` object,

Reply:
it's not clear-cut
https://stackoverflow.com/a/4557763

I'd leave it in.

Reply:
I agree.

Also, considering this is for a library meant for the web, it makes sense to emulate what *URLSearchParams* do in the same situation.

<img width="315" alt="image" src="https://github.com/TanStack/router/assets/33615041/b024f202-dedb-425c-90d5-c8f9a55817a7">